clients/src/main/java/org/apache/kafka/common/record/FileRecords.java

```diff
@@ -18,109 +18,465 @@
+                        shallowEntry.offset(), targetTimestamp));
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Return the largest timestamp of the messages after a given position in this file message set.
+     * @param startingPosition The starting position.
+     * @return The largest timestamp of the messages after the given position.
+     */
+    public TimestampAndOffset largestTimestampAfter(int startingPosition) {
+        long maxTimestamp = Record.NO_TIMESTAMP;
+        long offsetOfMaxTimestamp = -1L;
+
+        Iterator<FileChannelLogEntry> shallowIterator = shallowEntriesFrom(startingPosition);
+        while (shallowIterator.hasNext()) {
+            LogEntry shallowEntry = shallowIterator.next();
+            long timestamp = shallowEntry.record().timestamp();
+            if (timestamp > maxTimestamp) {
+                maxTimestamp = timestamp;
+                offsetOfMaxTimestamp = shallowEntry.offset();
+            }
         }
+        return new TimestampAndOffset(maxTimestamp, offsetOfMaxTimestamp);
     }
 
     @Override
-    public RecordsIterator iterator() {
-        return new RecordsIterator(new FileLogInputStream(channel, start, end), false);
+    public boolean hasMatchingShallowMagic(byte magic) {
+        Iterator<FileChannelLogEntry> iterator = shallowEntriesFrom(Integer.MAX_VALUE, start, false);
+        while (iterator.hasNext())
+            if (iterator.next().magic() != magic)
+                return false;
+        return true;
     }
 
-    private static class FileLogInputStream implements LogInputStream {
-        private long position;
-        protected final long end;
-        protected final FileChannel channel;
-        private final ByteBuffer logHeaderBuffer = ByteBuffer.allocate(Records.LOG_OVERHEAD);
+    @Override
+    public Iterator<FileChannelLogEntry> shallowIterator() {
+        return shallowEntriesFrom(start);
+    }
+
+    public Iterator<FileChannelLogEntry> shallowEntries(int maxMessageSize) {
+        return shallowEntriesFrom(maxMessageSize, start, true);
+    }
+
+    private Iterator<FileChannelLogEntry> shallowEntriesFrom(long start) {
+        return shallowEntriesFrom(Integer.MAX_VALUE, start, true);
+    }
+
+    private Iterator<FileChannelLogEntry> shallowEntriesFrom(int maxMessageSize, long start, boolean eagerLoadRecords) {
+        final long end;
+        if (isSlice)
+            end = this.end;
+        else
+            end = this.sizeInBytes();
+        FileLogInputStream inputStream = new FileLogInputStream(channel, maxMessageSize, start, end, eagerLoadRecords);
+        return RecordsIterator.shallowIterator(inputStream);
+    }
+
+    @Override
+    public Iterator<LogEntry> deepIterator() {
+        final long end;
+        if (isSlice)
+            end = this.end;
+        else
+            end = this.sizeInBytes();
+        FileLogInputStream inputStream = new FileLogInputStream(channel, Integer.MAX_VALUE, start, end, true);
```

Comment:
I followed what the current code did. Previously deep iteration was done here in `ByteBufferMessageSet.deepIterator`, which has no check for max message size.